docs/design/sql/10-distributed-scan.md

```diff
@@ -0,0 +1,146 @@
+
+If there are no indexes on the table, the direct scan is chosen and no further optimization is performed.    
+
+If there are indexes on the table, we analyze the predicate stored in the table of the `TableScan` operator. The entry point
+is `IndexResolver.createIndexScans`. 
+
+First, we split the predicate into conjunctive normal form (CNF). For example, the predicate `a=1 AND b=2` is split into 
+`a=1` and `b=2`, while the predicate `a=1 OR b=2` remains unchanged.
+
+Second, for every sub-predicate we find those that could be used by some index. Assume that `col` is a simple column
+expression, and `exp` is an expression that has only constants or parameters at leaves (i.e. it doesn't refer to other columns). 
+We use the following rules:
+1. `col = exp` can be used with `SORTED` and `HASH` indexes
+1. `col [comparison] exp` can be used with `SORTED` indexes
+1. `col IS NULL/TRUE/FALSE` can be treated as equality expression (with slightly different semantics for `NULL` values)
+1. `col = exp1 OR col = exp2` can be treated as union of two equality predicates 
+
+The result of this step is a map from column to candidate expressions that could be used with indexes. For example, 
+`a=1 AND b>2 AND b<4` is returned as: 
+```
+a -> [=1] 
+b -> [>2], [<4]
+```
+
+Next, we iterate over every index, and try to bind candidates to the index based on the index columns and index type.
+General rules are:
+1. `SORTED` index may use equality and comparison conditions, while `HASH` index may use only equality conditions
+1. Only a continuous prefix of index columns can be used. E.g. for the index `(a, b, c)` and the condition `a=1 AND c=2` we 
+can use only `a=1`
+1. All expressions in the prefix except for the last one must be equality conditions. E.g. for the index `(a, b)` and the 
+condition `a>1 AND b>2`, we can use only `a>1`   
+
+The result of this step is a map from index to the filter that should be used. For example, for the expression 
+`a=1 AND b>2 AND b<4`, and indexes on `(a, c)` and `(b)`, the result would be:
+```
+index(a, c) -> [a=1, NULL]
+index(b) -> [b>2 AND b<4]
+```
+
+At last, for every proposed index, we create a `MapIndexScanPhysicalRel` operator that is added to the planner search space.
+
+## 3 Local Execution
+
+Scan operators consist of three parts:
+1. Access path (direct scan, index scan)
+1. Optional filter
+1. Column list
+
+The execution proceeds as follows. First, an iterator over the underlying data is opened (record stores, index). Then every 
+returned record is evaluated against the optional filter. If record is not filtered out, it is converted to a `Row` instance
+based on the column list. The `Row` is added to the internal batch. When the batch reaches a certain size, or there are no
+more records, the batch of rows is returned to the parent operator. 
+
+For the direct map scan (`MapScanExec`), the iterator scans all local partitions one by one. For the index scan 
```

Comment:
Is there any specific reason why partitions are scanned one by one and not in parallel?

Reply:
We do not have operator parallelism yet, because this is a complicated topic. Parallelism increases total query resource consumption (e.g. memory buffers) and decrease total system throughput under load. We will have in the future, but after the careful design and prototype is ready.